### PR TITLE
[@types/js-yaml] Replace `object` with `Record`  in `load` and allow generic

### PR DESCRIPTION
**Pull request by joebeachjoebeach**
As I'm changing a type definition, here's the source code of the function whose definition I'm updating: https://github.com/nodeca/js-yaml/blob/ba3460eb9d3e4478edcbc29edabe17c2157fc9ce/lib/loader.js#L1713-L1723

There are two essential changes:

### 1. Removing `object` as a possible return type of `load`
`object` is something which has no properties on it; it's useful as a function parameter in some cases, but it is a hindrance as a return value.

Imagine writing this code with the current types for this package:
```ts
const myYaml = yaml.load(contents);
if (typeof myYaml !== object || myYaml === null) {
    // handle this case
}

const myValue = myYaml.myKey ?? "default value";

doThingWith(myValue);
```
This code looks pretty type safe, but TS throws an error at `myYaml.myKey` — `myYaml` is of type `object`, which has no keys on it. Even though we're providing a default value, we can't even try accessing that key. Our only option is to use a type assertion (e.g. `const myYaml = yaml.load(contents) as Record<string, unknown>`), which is seen as a bad practice in many code bases.

**I'm replacing `object` with `unknown[] | Record<string, unknown>`**, which matches the possible top-level types of YAML, being essentially an array or a dictionary-like structure.

### 2. Allowing for a generic in `load` in order to clarify the expected return type
Similarly to the types for DOM operations like `querySelector`, which accept generics to specify their return types (e.g. `document.querySelector<HTMLButtonElement>(".my-button")`), I've made `yaml.load` accept a generic that you can fill in with the shape of the data you expect to be returned.

### CODE REVIEW

File: types/js-yaml/index.d.ts [hunk 1]
```diff
@@ -9,7 +9,12 @@
 
 export as namespace jsyaml;
 
-export function load(str: string, opts?: LoadOptions): object | string | number | null | undefined;
+// tslint:disable-next-line:no-unnecessary-generics
+export function load<T = unknown[] | Record<string, unknown> | string | number>(
```
Comment (sandersn) [inline types/js-yaml/index.d.ts:13]: In this case, a generic isn't any safer than an assertion. It's basically an alternate, safer-looking syntax for the same thing.

**Review by sandersn — CHANGES_REQUESTED** [review]
The switch to `Record<string, unknown>` makes sense to me personally, but it does allow you to skip asserting the type of `load`'s result. I want to hear from the js-yaml maintainers whether requiring an assertion is what they intended.

File: types/js-yaml/index.d.ts (hunk 1 above)
Comment (joebeachjoebeach) [reply types/js-yaml/index.d.ts:13]: Fair enough. I'll remove it.

**Comment by joebeachjoebeach** [comment]
@sandersn thanks for the review! I've removed the generic.

> but it does allow you to skip asserting the type of `load`'s result

Just checking I understand — is the idea here that the author may have intended for it to be necessary to do something like `const yaml = load(myStr) as MyType`? and the phrase "skip asserting" it would mean that assertion is no longer necessary? Genuinely curious as a relative newcomer to TS — if I'm reading that correctly, what would be the argument in favor of a required assertion?

**Comment by sandersn** [comment]
Yes, the current types are stricter in that they require `const yaml = load(myStr) as MyType` before you can write `yaml.x`, whereas `Record<string, unknown>` would allow `yaml.x` without `as MyType`. That's more convenient, although `x`'s type is then unknown and has to be checked. It's a tradeoff.

Having thought more about it, I prefer the current code because you either have to write an unchecked assertion (`as MyType`) or pass the result into a checking function (like `declaration function isMyType(o: object): o is MyType`). Either way you type the whole thing. The Record just gives you one property access for free, but you still have to check the type of each property you access.

**Comment by joebeachjoebeach** [comment]
Ah, OK. This is making sense to me now; thanks for clarifying. I'll close this PR.

**Pull request closed without merging**